File: frontend/src/api/firmware.ts
import apiClient from "./client";

export interface FirmwareVersionRead {
  id: string;
  device_id: string;
  version: string;
  vendor_label: string;
  model: string | null;
  build: string | null;
  read_at: string;
  read_method: string;
}

export interface FirmwareCVERead {
  id: string;
  cve_id: string;
  vendor: string;
  product: string;
  affected_versions: Record<string, unknown>;
  cvss_v3: number | null;
  cvss_v2: number | null;
  severity: string;
  description: string;
  published_at: string | null;
  nvd_url: string;
  synced_at: string;
}

export interface FirmwareVulnRead {
  id: string;
  device_id: string;
  cve_id: string;
  device_version: string;
  detected_at: string;
  status: string;
  accepted_by: string | null;
  accepted_reason: string | null;
  patched_at: string | null;
  cve: FirmwareCVERead | null;
}

export interface DeviceFirmwareSummary {
  device_id: string;
  current_version: string | null;
  last_read_at: string | null;
  open_cves: number;
  critical_cves: number;
  high_cves: number;
  worst_severity: string;
}

export interface FirmwareRiskSummary {
  devices_with_vulns: number;
  total_open_cves: number;
  critical_cves: number;
  high_cves: number;
  top_affected: DeviceFirmwareSummary[];
}

export const firmwareApi = {
  getSummary: (deviceId: string) =>
    apiClient.get<DeviceFirmwareSummary>(`/devices/${deviceId}/firmware/summary`).then(r => r.data),

  getVersions: (deviceId: string) =>
    apiClient.get<FirmwareVersionRead[]>(`/devices/${deviceId}/firmware/versions`).then(r => r.data),

  getVulnerabilities: (deviceId: string, status = "open") =>
    apiClient.get<FirmwareVulnRead[]>(`/devices/${deviceId}/firmware/vulnerabilities`, { params: { status } }).then(r => r.data),

  triggerRefresh: (deviceId: string) =>
    apiClient.post<{ task_id: string; status: string }>(`/devices/${deviceId}/firmware/refresh`).then(r => r.data),

  acceptRisk: (vulnId: string, reason: string) =>
    apiClient.patch(`/firmware/vulnerabilities/${vulnId}/accept`, { reason }).then(r => r.data),

  getRiskSummary: () =>
    apiClient.get<FirmwareRiskSummary>("/firmware/risk-summary").then(r => r.data),

  listCVEs: (vendor?: string, severity?: string) =>
    apiClient.get<FirmwareCVERead[]>("/firmware/cves", { params: { vendor, severity } }).then(r => r.data),
};

File: backend/app/schemas/firmware.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel


class FirmwareVersionRead(BaseModel):
    id: UUID
    device_id: UUID
    version: str
    vendor_label: str
    model: str | None
    build: str | None
    read_at: datetime
    read_method: str

    model_config = {"from_attributes": True}


class FirmwareCVERead(BaseModel):
    id: UUID
    cve_id: str
    vendor: str
    product: str
    affected_versions: dict
    cvss_v3: float | None
    cvss_v2: float | None
    severity: str
    description: str
    published_at: datetime | None
    nvd_url: str
    synced_at: datetime

    model_config = {"from_attributes": True}


class FirmwareVulnRead(BaseModel):
    id: UUID
    device_id: UUID
    cve_id: str
    device_version: str
    detected_at: datetime
    status: str
    accepted_by: UUID | None
    accepted_reason: str | None
    patched_at: datetime | None
    cve: FirmwareCVERead | None = None

    model_config = {"from_attributes": True}


class FirmwareVulnAccept(BaseModel):
    reason: str


class DeviceFirmwareSummary(BaseModel):
    device_id: UUID
    current_version: str | None
    last_read_at: datetime | None
    open_cves: int
    critical_cves: int
    high_cves: int
    worst_severity: str


class FirmwareRiskSummary(BaseModel):
    devices_with_vulns: int
    total_open_cves: int
    critical_cves: int
    high_cves: int
    top_affected: list[DeviceFirmwareSummary]
